### ISSUE TITLE
OAuth: Support groups overage claim in AzureAD

### ISSUE THREAD

**Issue by wiegandf** [grafana/grafana#28457]
**What happened**:
I want to use Grafana with AzureAD authentication. Only people from certain groups should be allowed to login. Therefore, I use auth.azuread.allowed_groups with object id of just a single group.

For some people of this group it works, for others it doesn't work. 

In case it works, I can see 
```
{"logger":"oauth","lvl":"dbug","msg":"OAuthLogin Got token","t":"2020-10-22T08:59:45.011077157Z","token":{"access_token":"valid_token","token_type":"Bearer","expiry":"2020-10-22T09:59:44.010987224Z"}}
{"logger":"oauth","lvl":"dbug","msg":"OAuthLogin got user info","t":"2020-10-22T08:59:45.012054798Z","userInfo":{"Id":"abc","Name":"b, a","Email":"a@b.com,"Login":"a@b.com","Company":"","Role":"Viewer","Groups":["abc", "cde"]}}
```

For people it doesn't work:
```
{"logger":"oauth","lvl":"dbug","msg":"OAuthLogin Got token","t":"2020-10-22T09:08:25.589783186Z","token":{"access_token":"valid_token","token_type":"Bearer","expiry":"2020-10-22T10:08:24.589722595Z"}}
{"logger":"context","lvl":"eror","msg":"User not a member of one of the required groups","orgId":0,"t":"2020-10-22T09:08:25.590050571Z","uname":"","userId":0}
```

**What you expected to happen**:
All people from allowed group should have access. 

**How to reproduce it (as minimally and precisely as possible)**:
Not figured out the root cause, so it's difficult to say.

**Anything else we need to know?**:

- It looks like users with more than 75 groups are not able to login (or with membership of a certain group?). 
- Dex is using the same azure application and groups and it's working for all people.
- The jwt tokens for both cases look similar (same fields - all non-empty)

**Environment**:
- Grafana version: 7.1.5

**Comment by marefr** [comment]
> The jwt tokens for both cases look similar (same fields - all non-empty)

Can you please include examples of those? And please replace any sensitive information. Thanks

**Comment by wiegandf** [comment]
Sure, @marefr.

This one is working:
```
{
  "typ": "JWT",
  "nonce": "3R6yB59_N5jJFpOZfRodddQfvo5HI_cDhJ_Y29UjdWg",
  "alg": "RS256",
  "x5t": "kg2LYs2T0CTjIfj4rt6JIynen38",
  "kid": "kg2LYs2T0CTjIfj4rt6JIynen38"
}.{
  "aud": "00000003-0000-0000-c000-000000000000",
  "iss": "https://sts.windows.net/42f7676c-f455-423c-82f6-dc2d99791af7/",
  "iat": 1603717083,
  "nbf": 1603717083,
  "exp": 1603720983,
  "acct": 0,
  "acr": "1",
  "aio": "AVQAq/8RAAAAgcZgY6zZYrOa9C+1C3TqORYmorDmgeIxktwwGeJ+9wednWITxTt0Pb5LR9OqXevYkxd3Ydr0PRgWpsGQXCVvqyrnTl74xzIe5A53R1sWsa4=",
  "amr": [
    "rsa",
    "mfa"
  ],
  "app_displayname": "app_name",
  "appid": "8053e688-b290-4539-83dc-f9e05b443af1",
  "appidacr": "1",
  "family_name": "b",
  "given_name": "a",
  "idtyp": "user",
  "in_corp": "true",
  "ipaddr": "x.x.x.x",
  "name": "b, a",
  "oid": "f1a022ac-f2a0-4b9d-802b-dce6a81788d3",
  "onprem_sid": "S-1-5-21-74642-3284969411-2123768488-1900491",
  "platf": "5",
  "puid": "100320007D8C91B6",
  "rh": "0.AQkAbGf3QlX0PEKC9twtmXka94jmU4CQsjlFg9z54FtEOvEJABo.",
  "scp": "Directory.Read.All User.Read profile openid email",
  "sub": "hlKwg8ux0T0Gjnhfor1f8_86KUiyGSImAGJpqiECVi4",
  "tenant_region_scope": "EU",
  "tid": "42f7676c-f455-423c-82f6-dc2d99791af7",
  "unique_name": "a.b@org.com",
  "upn": "a.b@org.com",
  "uti": "tsQ7zOkYFEeJp55c3v17AA",
  "ver": "1.0",
  "wids": [
    "b79fbf4d-3ef9-4689-8143-76b194e85509"
  ],
  "xms_st": {
    "sub": "V_kN7bLpE1L277qerd72oBT__TWwMghDL2cZSafbdHk"
  },
  "xms_tcdt": 1371838550
}.[Signature]
```

This one is not working
```
{
  "typ": "JWT",
  "nonce": "U8Pp2Qc1sIZNYA0_F7gQSU9dEvlIDJUXr2T_b0A5hx0",
  "alg": "RS256",
  "x5t": "kg2LYs2T0CTjIfj4rt6JIynen38",
  "kid": "kg2LYs2T0CTjIfj4rt6JIynen38"
}.{
  "aud": "00000003-0000-0000-c000-000000000000",
  "iss": "https://sts.windows.net/42f7676c-f455-423c-82f6-dc2d99791af7/",
  "iat": 1603717370,
  "nbf": 1603717370,
  "exp": 1603721270,
  "acct": 0,
  "acr": "1",
  "aio": "AVQAq/8RAAAABvVYYS2g5Xu8S0XKdeQ4Wz3+KnoNYml/pIzzaZJVauSkqvOJv26iSD4XR9rzZrV8ChHMo/OWcAjC7gHtkMpMFhb3fSUMgZyA31R9nZC5HK4=",
  "amr": [
    "rsa",
    "mfa"
  ],
  "app_displayname": "app_name",
  "appid": "8053e688-b290-4539-83dc-f9e05b443af1",
  "appidacr": "1",
  "family_name": "d",
  "given_name": "c",
  "idtyp": "user",
  "in_corp": "true",
  "ipaddr": "x.x.x.x",
  "name": "d, c",
  "oid": "50cd88fc-76f9-402b-a82f-3fedd3994c1a",
  "onprem_sid": "S-1-5-21-74642-3284969411-2123768488-498234",
  "platf": "5",
  "puid": "10033FFF944A28ED",
  "rh": "0.AQkAbGf3QlX0PEKC9twtmXka94jmU4CQsjlFg9z54FtEOvEJAD4.",
  "scp": "Directory.Read.All User.Read profile openid email",
  "sub": "vMXA7f-N2A3v-PHll3ffvpyeenzopIV4V0oPEzzM6DU",
  "tenant_region_scope": "EU",
  "tid": "42f7676c-f455-423c-82f6-dc2d99791af7",
  "unique_name": "c.d@org.com",
  "upn": "c.d@org.com",
  "uti": "rL7tfSP850u_FYAD64FMAA",
  "ver": "1.0",
  "wids": [
    "b79fbf4d-3ef9-4689-8143-76b194e85509"
  ],
  "xms_st": {
    "sub": "dPwIupoYmZ8B5P7qkjZ5hiGpUTm89_VVqa_AP6wiKXo"
  },
  "xms_tcdt": 1371838550
}.[Signature]
```

**Comment by marefr** [comment]
Hmm I don't see any groups property in those payloads? Can you please include your Grafana azuread configuration (please replace sensitive information).

**Comment by wiegandf** [comment]
/etc/grafana/grafana.ini:
```
[auth]
disable_login_form = true
oauth_auto_login = true
[auth.azuread]
allowed_groups = fcac86b2-08ea-4d29-839d-441fc4725f3e
auth_url = https://login.microsoftonline.com/42f7676c-f455-423c-82f6-dc2d99791af7/oauth2/v2.0/authorize
client_id = <id>
client_secret = <secret>
enabled = true
token_url = https://login.microsoftonline.com/42f7676c-f455-423c-82f6-dc2d99791af7/oauth2/v2.0/token
[auth.basic]
enabled = false
```

**Comment by marefr** [comment]
Thanks. In the payloads you posted earlier I don't see a reference to `fcac86b2-08ea-4d29-839d-441fc4725f3e`. Based on the code I would expect a property "groups" in the payload. 

Have you enabled group attributes as explained in [here](https://grafana.com/docs/grafana/latest/auth/azuread/#configure-allowed-groups-and-domains)?

**Comment by wiegandf** [comment]
That's what I would have expected as well. Interestingly, it still works for about half the people in the allowed group (without the groups property in the jwt payload).
 
I followed the documentation and `"groupMembershipClaims": "ApplicationGroup"` is set. Actually, I also followed https://github.com/grafana/grafana/pull/23465 because otherwise it didn't work at all.

I'll try to add more debug logs to 
https://github.com/grafana/grafana/blob/7afdfd2ef46c54ccbec275b45b340def4846d112/pkg/login/social/azuread_oauth.go#L127

**Comment by marefr** [comment]
Maybe @alexanderzobnin can help bring some clarity to this as well?

**Comment by alexanderzobnin** [comment]
@wiegandf Hi! Could you post content of the `id_token`, because it contains groups? The tokens above are access tokens. You can find encoded id token in the same grafana log message that contains access token.

**Comment by wiegandf** [comment]
So for both cases (working & non-working) I can't see the id_token field.

Working case:
```
{"logger":"oauth","lvl":"dbug","msg":"OAuthLogin Got token","t":"2020-11-03T07:12:12.661492091Z","token":{"access_token":"<see_above>","token_type":"Bearer","expiry":"2020-11-03T08:12:11.661393877Z"}}
{"logger":"oauth","lvl":"dbug","msg":"OAuthLogin got user info","t":"2020-11-03T07:12:12.661997426Z","userInfo":{"Id":"<user_id>","Name":"b, a","Email":"a.b@c.com","Login":"a.b@c.com","Company":"","Role":"Viewer","Groups":<list_of_group_ids}}
```
Non-working case:
```
{"logger":"oauth","lvl":"dbug","msg":"OAuthLogin Got token","t":"2020-11-03T08:27:44.823121667Z","token":{"access_token":"<see_above>","token_type":"Bearer","expiry":"2020-11-03T09:27:43.823087729Z"}}
{"logger":"context","lvl":"eror","msg":"User not a member of one of the required groups","orgId":0,"t":"2020-11-03T08:27:44.823346096Z","uname":"","userId":0}
```

**Comment by alexanderzobnin** [comment]
Message looks like (id_token is in the raw data):
```
DBUG[11-03|14:02:24] OAuthLogin Got token    logger=oauth token="&{AccessToken:eyJ0eXA... TokenType:Bearer RefreshToken: Expiry:2020-11-03 15:02:23.386054227 +0300 MSK m=+17952.565242938 raw:map[access_token:eyJ0eXA... expires_in:3599 ext_expires_in:3599 id_token:eyJ0eXA... scope:email openid profile token_type:Bearer]}"
```

**Comment by wiegandf** [comment]
Thank you very much - this pointed me to reason it's not working. Unfortunately, the id_token wasn't logged with log format json which I would not have expected.

So, for the non-working case, id_token consists of a groups overage claim (https://docs.microsoft.com/de-de/azure/active-directory/develop/id-tokens#groups-overage-claim).

**Comment by marefr** [comment]
Thanks. I re-labeled this as a feature request since this was not something thought of when implemented. We'll keep it open to see if there will be more interest/activity in it,

**Comment by zachwalton** [comment]
Hopefully a comment is the right way to signal interest / activity :)

This is preventing me from using Grafana with GitHub Enterprise oauth2 authentication in my (admittedly unusually large) organization where I'm a member of 82 teams.

(trying to leave some as a workaround)

**Comment by Jguer** [comment]
This should be now be available with accompanying fixes in the latest version of 8.4, 8.5 and 9.0. 

https://github.com/grafana/grafana/pull/43470
https://github.com/grafana/grafana/pull/49506
